Regular sync with upstream DS

Quentin-Anthony opened the pull request.

lekurile committed:
Fix Meta Tensor checkpoint load for OPT models (#2990)

This PR fixes Meta Tensor checkpoint loading for OPT models where the SD keys start with `model.`.

adammoody committed:
ckpt: create directories in checkpoint_engine (#2988)

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

tohtana committed:
Fix buffer size for pipeline parallel and communication schedule (#2862)

* fix buffer size for pipeline parallel (#2800)

* improve explanation of buffer size for pipeline parallelism

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

* fix format of comment

---------

Co-authored-by: Jeff Rasley <jerasley@microsoft.com>
Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>
Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

jeffra committed:
[docs] add new paper to readme/docs (#3018)

Co-authored-by: Zhewei Yao <zheweiyao@gmail.com>

stas00 committed:
Update torch_checkpoint_engine.py (#3019)

jomayeri committed:
adding attribute checks for bf opt with zero (#3022)

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

stas00 committed:
[logger] implement warning_once (#3021)

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

jomayeri committed:
Convert model parameters from generator to list. (#3017)

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

Quentin-Anthony committed:
Improve loss overflow logs (#3008)

* Improve overflow logs

* Trigger CI

---------

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

jeffra committed:
update email info

satpalsr committed:
Fix Broken Links (#3048)

Quentin-Anthony committed:
[docs] add MCR-DL paper to readme/docs (#3066)

loadams committed:
Several fixes to unblock CI (#3047)

Co-authored-by: Michael Wyatt <michaelwyatt@microsoft.com>
Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

molly-smith committed:
Assert mp_size is factor of model dimensions (#2891)

Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

jeffra committed:
[CI] follow-up fixes (#3072)

mzusman committed:
fix return prev key and value , added strides to from_blob (#2828)

Co-authored-by: Mor Zusman <morz@ai21.com>
Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>
Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

molly-smith committed:
Remove bf16 from inference config dtye enum (#3010)

Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

cmikeh2 committed:
Softmax Scheduling Cleanup (#3046)

Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

FreyaRao committed:
Fix nebula in save_16bit_model issue (#3023)

Co-authored-by: Qinghuan Rao <qinghuanrao@microsoft.com>

satpalsr committed:
allow list (#3042)

Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

mrwyattii committed:
Goodbye Torch 1.8 (#3082)

* bump torch18 -> torch19
* fix gptj

---------

Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

tjruwase committed:
Empty ZeRO3 partition cache (#3060)

Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

delock committed:
pre-commit check for torch.cuda in code (#2981)

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

loadams committed:
Move cuda check into utils (#3074)

Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

jeffra committed:
update formatter version and style settings (#3098)

Quentin-Anthony committed:
Fix comms benchmark import issues and support MPI/slurm launching (#2932)

* Fix benchmark import issues and support MPI launching with pure torch.dist

* Formatting

* Update comms benchmark README

* Formatting

* Added better error handling and support MPI torch.dist backend

* Update formatting versions

* Formatting again

* Trigger CI

---------

Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

mrwyattii committed:
Disable Stage 1&2 CPUAdam pathways (#3097)

* disable CPUAdam pathways in optimizer copy/step

* Update stage_1_and_2.py

---------

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

mayank31398 committed:
♻️ replace deprecated functions for communication (#2995)

* :hankey: drop dead code

* :recycle: replace has_all_gather_base with has_all_gather_into_tensor

* :recycle: remove deprecated _all_gather_base

* :recycle: remove deprecated _reduce_scatter_base

* :art: reformat files

* :wrench: fix _six

* Trigger CI

* Trigger CI

* Trigger CI

* :art: formatting

* incorporate suggestion

* incorporate suggestion

---------

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>
Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

tjruwase committed:
Make fp32 default communication data type (#2970)

* Make fp32 default communication data type

* Fix asserts

mrwyattii committed:
Update DeepSpeed copyright license to Apache 2.0 (#3111)

Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

mrwyattii committed:
added full apache license (#3119)

yaozhewei committed:
VL MoE Blog (#3120)

Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

jeffra committed:
[docs] fix blog links

lekurile committed:
Update SD triton version in requirements-sd.txt (#3135)

tjruwase committed:
Fix launch issue (#3137)

mrwyattii committed:
fix CI badges (#3138)

molly-smith committed:
Optimize Softmax Kernel (#3112)

* Simplify kernel

* Coalesce memory attempt 1. Logits divergence.

* Logits fix?

* sync after every global mem access

* template on iterations. Down to 8.3% cuda time for 8k tokens

* Up to 64 iterations

* Add alibi/mask check

* fp32

* Revert builder.py

* naming. precommit

* Revert "naming. precommit"

This reverts commit 150eb7d96b6084190265b440739317216992bd82.

* naming. spacing

* Spacing. simplify checks

* remove bsyncs

* missed bsyncs

* precommit

tjruwase committed:
Use generic O_DIRECT (#3115)

sywangyi committed:
Enable autoTP for bloom (#3035)

Signed-off-by: Wang, Yi A <yi.a.wang@intel.com>
Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

stas00 committed:
Update engine.py (#2826)

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

stas00 committed:
[ci] `nv-transformers-v100` - use the same torch version as transformers CI (#3096)

Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>
Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

loadams committed:
Update skip on torch in tests (#3136)

* Replace old torch version checks with existing function

* Clean up formatting

EeyoreLee committed:
fix example symlink about DeepSpeed+AzureML (#3127)

Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>

VHellendoorn committed:
Remove Extra Bracket (#3101)

This was breaking the link to the Pyramid MoE paper on Arxiv

loadams committed:
Update megatron GPT2Model

loadams committed:
Revert "Update megatron GPT2Model"

This reverts commit 1ec34e54de6bf07f85802c1ae104a876f6e65cf9.

ShijieZZZZ committed:
Recover shared parameters (#3033)

* submit changes

* update format

* fix fomrat

* revert

* test

* add top

* treat z1 as z2

* revert

---------

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

molly-smith committed:
Fix for Diffusers 0.14.0 (#3142)

cross attention kwargs and vae config for diffusers 0.14.0

mrwyattii committed:
Fix copyright check, add copyright replace script (#3141)

* fix copyright script and add replace-copyright script

goodship1 committed:
Update curriculum-learning.md (#3031)

Co-authored-by: Jeff Rasley <jerasley@microsoft.com>
Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

mrwyattii committed:
remove benchmarks (now in DSE) and add links (#3157)

Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

xiexbing committed:
fixing a bug in CPU Adam and Adagrad (#3109)

Co-authored-by: Bing Xie <bingxie@BINGHYPC014.redmond.corp.microsoft.com>
Co-authored-by: Shaden Smith <Shaden.Smith@microsoft.com>
Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>
Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

adammoody committed:
op_builder: conditionally update relative path for hip compiled files (#3095)

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>
Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

loadams committed:
Test with triton 2

loadams committed:
Revert "Test with triton 2"

This reverts commit 0f8b5da6ae37b474259dd10164ffbd9f36ebb598.

tjruwase committed:
zero.Init() should pin params in GPU memory as requested (#2953)

* Persist params in zero.Init

* Disable debug prints

* Formatting

* Avoid offloading persisted params

* Simplify world_size=1

* Formatting

* Remove pdb

* Restructure

* Formating

* Formatting

* Apply persistence only if ds_config available

* Fix typo

* add util function for getting pydantic config default values

---------

Co-authored-by: Michael Wyatt <michaelwyatt@microsoft.com>

guoyejun committed:
deepspeed/runtime/utils.py: reset_peak_memory_stats when empty cache (#2803)

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

awan-10 committed:
Add DeepSpeed-Chat Blogpost (#3185)

Co-authored-by: Samyam Rajbhandari <samyamr@microsoft.com>
Co-authored-by: Reza Yazdani <reyazda@microsoft.com>
Co-authored-by: Samyam Rajbhandari <samyamr@microsoft.com>
Co-authored-by: Conglong Li <conglong.li@gmail.com>
Co-authored-by: Jeff Rasley <jerasley@microsoft.com>
Co-authored-by: Zhewei Yao <zheweiyao@gmail.com>

awan-10 committed:
add run command for 13b (#3187)

awan-10 committed:
add news item. (#3188)

tjruwase committed:
DeepSpeed Chat (#3186)

Co-authored-by: Reza Yazdani <44502768+RezaYazdaniAminabadi@users.noreply.github.com>
Co-authored-by: yaozhewei <zheweiy@berkeley.edu>
Co-authored-by: Ammar Ahmad Awan <ammar.awan@microsoft.com>
Co-authored-by: Jeff Rasley <jerasley@microsoft.com>
Co-authored-by: Connor Holmes <connorholmes@microsoft.com>
Co-authored-by: Lok Chand Koppaka <lokoppak@microsoft.com>
Co-authored-by: Masahiro Tanaka <81312776+tohtana@users.noreply.github.com>
Co-authored-by: Michael Wyatt <michaelwyatt@microsoft.com>

tohtana committed:
fix references to figures (#3189)

Co-authored-by: Michael Wyatt <michaelwyatt@microsoft.com>

zhouzaida committed:
fix typo (#3183)

dawei-wang committed:
Fix typo (#3164)

Fix microsoft/DeepSpeed#3163

yaozhewei committed:
Chatgpt chinese blog (#3193)

Co-authored-by: Ammar Ahmad Awan <ammar.awan@microsoft.com>

tohtana committed:
Add Japanese version of ChatGPT-like pipeline blog (#3194)


Co-authored-by: Conglong Li <conglong.li@gmail.com>

conglongli committed:
fix figure (#3199)

Commit:
feat: Add support for `NamedTuple` when sharding parameters [#3029] (#3037)

* feat: Add support for `NamedTuple` when sharding parameters [#3029]

* Formatting

---------

Co-authored-by: Alexander van Eck <alexander.vaneck@paige.ai>
Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

conglongli committed:
fix license badge (#3200)

loadams committed:
Update AMD workflows (#3179)

* Update AMD workflows

* Update MI200 test flow to use torch latest

* Update tolerances to values that pass (will fix before completing PR)

* Revert chyanges to atol

* Rename workflows

* Fix CI badges

delock committed:
[CPU support] Optionally bind each rank to different cores on host (#2881)

* add fallback path for kernels used in megatron

* temporary numactl WA for SPR 56core

* adapt core allocation according to number of ranks

* add switch to turn on numactl

* detect number of cores on the system

* allow select a subset of the cores on the system to bind

* remove unneeded changes

* use current_env to set OMP_NUM_THREADS in subprocess

* add test for ds_arguments

* change --bind_cores_to_rank option to store_true

* add test for parse_range_list

* add comment for parse range list

* add test for parse range list, rewrite parse_range_list

* fix format error

* fix format

* add -m parameter to numactl when necessary

* Check KMP_AFFINITY to avoid conflict with numactl

* fix format

* negative case for parse_range_list

* detect whether numactl is installed before use numactl to bind cores

* check numactl with package manager of distro

---------

Co-authored-by: sdp <sdp@aia-sdp-spr-108864.jf.intel.com>
Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

jeffra committed:
bump to 0.9.1

mrwyattii committed:
change pip install of deepspeed to 0.9.0 (#3216)

mrwyattii committed:
Update DeepSpeed-Chat docs with latest changes to scripts (#3219)

* update docs to reflect changes in deepspeed-chat training script

* add blogs to ignored changes in unit tests

tohtana committed:
Nested zero.Init() and dynamically defined model class (#2989)

* support nesting zero.Init() and dynamically defined module

* throw an error if model class defined in zero.Init is not wrapped

* fix check on new classes that are not wrapped in zero.Init()

* add tests of nesting zero.Init() and dynamically defined classes

* fix tests for zero.Init

* fix style

---------

Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>
Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

loadams committed:
Update torch version check in building sparse_attn (#3152)

* Update torch version check in building sparse_attn

* Update triton check

* Fix error message

* Format fixes

* Test with triton 2

* Change requirements back

* fix to just prevent 2.0.0

* Fix formatting

---------

Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

mrwyattii committed:
Create deepspeed_chat_bug_report.md

mrwyattii committed:
fix for SD afte DeepSpeed-chat merge (#3218)

3xdti committed:
[update] reference in cifar-10.md (#3212)

Co-authored-by: Michael Wyatt <michaelwyatt@microsoft.com>

stas00 committed:
[fp16/doc] correct initial_scale_power default value (#3275)

Borda committed:
update link to PL docs (#3237)

update link to PL docs talking about DeepSpeed integration

eltociear committed:
fix typo in autotuner.py (#3269)

resutls -> results

Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>

HeyangQin committed:
improving int4 asymmetric quantization accuracy (#3190)

* Fixes for asymmetric quantization

* addtional offset to further improve accuracy

* put the 0.5 into offset rather than applying it later

* update unit test for quantization

* fix format

* attempt to fix format

---------

Co-authored-by: Connor Holmes <connorholmes@microsoft.com>
Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

digger-yu committed:
Update install.sh (#3270)

Optimization Code
1. Use #!/usr/bin/env bash instead of #!/bin/bash to make the script more portable.
2. Use rm -rf instead of rm -r to remove directories recursively.

Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>
Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

mrwyattii committed:
Fix cupy install version detection (#3276)

* updated cupy install

* do non-isolated pip install

* Update action.yml

bmedishe committed:
[ROCm] temporary workaround till __double2half support enabled in HIP (#3236)

* temporary WAR workaround till __double2half support enabled in HIP

* workaround only for hipcc

---------

Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>

loadams committed:
Fix pydantic and autodoc_pydantic (#3290)

digger-yu committed:
Update README.md (#3282)

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

jinzhen-lin committed:
remove `torch.cuda.is_available()` check when compiling ops (#3085)

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>
Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>
Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

loadams committed:
Update MI200 workflow to install apex with changes from pip (#3294)

loadams committed:
Add pre-compiling ops test (#3277)

Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

digger-yu committed:
Update README.md (#3315)

Optimization Code
GitHub's image caching mechanism will cache images,Add a random number after the last modified link.so that every time you visit that link, the contributor's image will be refreshed in real time.

bobowwb committed:
Update Dockerfile (#3298)

line 98 should be
curl -O https://bootstrap.pypa.io/pip/3.6/get-pip.py && \
to avoid
#16 106.9 ERROR: This script does not work on Python 3.6 The minimum supported Python version is 3.7. Please use https://bootstrap.pypa.io/pip/3.6/get-pip.py instead.

Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>

tjruwase committed:
zero3 checkpoint frozen params (#3205)

* zero3 checkpoint frozen params

* Remove debug prints

* Move to cpu

* WIP

* WIP

* WIP

* Cleanup

* Cleanup

* Extend unit test for frozen params

* API fix

mrwyattii committed:
Fix for dist not being initialized when constructing main config (#3324)

* move dist init out of Engine

cmikeh2 committed:
Fix missing scale attributes for GPTJ (#3256)

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>
Co-authored-by: Michael Wyatt <michaelwyatt@microsoft.com>

cmikeh2 committed:
Explicitly check for OPT activation function (#3278)

Co-authored-by: Michael Wyatt <michaelwyatt@microsoft.com>
Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

jeffra committed:
bump to 0.9.2

mrwyattii committed:
Fix formatting (#3343)

* formatting

* fixing clang-format version

* update pre-commit URL

rraminen committed:
Hipify cooperative_groups headers (#3323)

Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>

molly-smith committed:
Diffusers 0.15.0 bug fix (#3345)

* diffusers 0.15.0 cross attention class check

* revert diffusers_attention.py

mrwyattii committed:
print default values (#3347)

dc3671 committed:
add bf16 cuda kernel support (#3092)

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

kobindra committed:
README.md: Update MosaicML docs link (#3344)

Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

adammoody committed:
hybrid_engine: check tuple size when fusing lora params (#3311)

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>
Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>

sywangyi committed:
fix mpich launcher issue in multi-node (#3078)

Signed-off-by: Wang, Yi A <yi.a.wang@intel.com>
Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

mrwyattii committed:
Update DS-Chat issue template (#3368)

* request log output

* add more details

conglongli committed:
add deepspeed chat blog links, add tags (#3369)

ShijieZZZZ committed:
Fix redundant shared_params in zero_to_fp32.py (#3149)

* submit changes

* update format

* fix fomrat

* revert

* test

* add top

* treat z1 as z2

* fix shared

* remove old changes

---------

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

Commit:
fixing default communication_data_type for bfloat16_enabled and docs (#3370)

Co-authored-by: Alexander Jipa <azzhipa@amazon.com>
Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>
Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>

molly-smith committed:
Auto TP Tutorial with T5 Example (#2962)

Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

zarzen committed:
[DRAFT] Tentative implementation of MiCS (#2964)

* include mics config and optimizer

* change private vars to public vars

so the child class can initialize these vars

* Port the init function from stage3

* adding a model test file for mics

* adopt to get_acceleartor api and fp16 group defrag

* WIP: porting mics modification to ms master

* WIP: included gradient all-reduce among replication groups

* WIP: ported hierarchical all gather part

did basic loss test on a simple MLP model

* [Bug fix] using the comm group attached on the param

* torch2.0 support

* remove print

* delegate wait op

* [Bug] fix naming

* adding doc string

* resolving recursive import

* fix formating, typo and license

* fix license and unit test error

---------

Co-authored-by: Ubuntu <ubuntu@ip-172-31-14-191.us-west-2.compute.internal>
Co-authored-by: Ubuntu <ubuntu@ip-172-31-7-70.us-west-2.compute.internal>
Co-authored-by: Zhen Zhang <zhzhn@amazon.com>
Co-authored-by: zhzhn <zhzhn@ip-10-2-57-114.us-west-2.compute.internal>

guoyejun committed:
stage_1_and_2.py: do gradient scale only for fp16 (#3166)

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

BacharL committed:
Fix memory leak in zero2 contiguous gradients (#3306)

No usage of extra_large_param_to_reduce if contiguous_gradients is False.
It keeps reference of the param for the lifetime of the application.

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>
Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>

jeffra committed:
[docs] add twitter badge to readme

jeffra committed:
remove megatron-lm, no longer pip installable (#3389)

* remove megatron-lm, no longer pip installable

* Add skips to tests that require megatron-lm and can't be run currently.

* formatting

* Formatting

---------

Co-authored-by: Logan Adams <loadams@microsoft.com>

BacharL committed:
Fix pipeline module evaluation when contiguous activation checkpointing is enabled (#3005)

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

goodship1 committed:
doc updates (#3415)

tohtana committed:
save tensors in context of memory_efficient_linear (#3413)

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

RezaYazdaniAminabadi committed:
Add HE support for the rest of model containers (#3191)

Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

loadams committed:
Update links (#3424)

nrailg committed:
Fix `PipelineEngine.eval_batch` result (#3316)

Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

cmikeh2 committed:
OPT Activation Function Hotfix (#3400)

Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>
Co-authored-by: Zhewei Yao <zheweiyao@gmail.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/EleutherAI/DeeperSpeed?pullRequest=50) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/EleutherAI/DeeperSpeed?pullRequest=50) before we can accept your contribution.<br/>**0** out of **19** committers have signed the CLA.<br/><br/>:x: molly-smith<br/>:x: dc3671<br/>:x: mrwyattii<br/>:x: adammoody<br/>:x: conglongli<br/>:x: sywangyi<br/>:x: kobindra<br/>:x: zarzen<br/>:x: guoyejun<br/>:x: hablb<br/>:x: clumsy<br/>:x: ShijieZZZZ<br/>:x: tohtana<br/>:x: nrailgun<br/>:x: loadams<br/>:x: RezaYazdaniAminabadi<br/>:x: goodship1<br/>:x: jeffra<br/>:x: cmikeh2<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/EleutherAI/DeeperSpeed?pullRequest=50) it.</sub>

jomayeri committed:
Add ZeRO 1 support to PP for BF16. (#3399)

* Add ZeRO 1 support to PP for BF16.

* Switching enum.

---------

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

stas00 committed:
[zero_to_fp32] fix shared param recovery (#3407)

* [zero_to_fp32] fix shared param recovery

* cleanup

* cleanup

* better naming

* not all params have ds_id it seems

---------

Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>
Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

jomayeri committed:
Adagrad support in ZeRO (#3401)

* Adding torch.optim.Adagrad

* adding adagrad for zero 1 2

* Adding Adagrad support to zero 3.

* Adding documentation and DeepSpeedCPUAdagrad to list.

---------

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

goodship1 committed:
Update 2020-09-09-sparse-attention.md (#3432)

Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>

jeffra committed:
bump to 0.9.3

jianan-gu committed:
Enable auto TP policy for llama model (#3170)

* Enable auto TP policy for llama model

* Update automatic-tensor-parallelism.md

---------

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>
Co-authored-by: Molly Smith <112220543+molly-smith@users.noreply.github.com>

mrwyattii committed:
add skip for mismatched cuda versions (#3436)

cmikeh2 committed:
Hybrid Engine Refactor and Llama Inference Support (#3425)

Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

sywangyi committed:
add sharded checkpoint loading for AutoTP path to reduce the peak mem… (#3102)

* add sharded checkpoint loading for AutoTP path to reduce the peak memory in initialization stage

Signed-off-by: Wang, Yi A <yi.a.wang@intel.com>

* fix gptj sharded checkpoint loading problem

Signed-off-by: Wang, Yi A <yi.a.wang@intel.com>

---------

Signed-off-by: Wang, Yi A <yi.a.wang@intel.com>
Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

YizhouZ committed:
launcher/multinode_runner.py: mapping env variables in running cmd for mpich runner (#3372)

Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>
Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

sywangyi committed:
[Docs] Update automatic-tensor-parallelism.md (#3198)

remove bloom from unsupported Models

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>
Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>

PabloEmidio committed:
Build: Update license in setup (#3484)

goodship1 committed:
Doc corrections  (#3435)

* Update 2020-09-09-sparse-attention.md

* Update MoQ-tutorial.md

---------

Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>

digger-yu committed:
Fix spelling errors in comments and documents  (#3486)

* fix some spelling error under doc/

* fix spelling error deepspeed/

---------

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>
Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>

luliyucoordinate committed:
fix spelling error (#3482)

Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>

yisustc committed:
fix a small type error on bf16+pp (#3441)

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>
Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>

digger-yu committed:
fix spelling error with deepspeed/ (#3494)

digger-yu committed:
fix spelling error with  docs/index.md (#3443)

* Update index.md

fix spelling error

* Update training.md

fix spelling error

---------

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

MrZhengXin committed:
delete the line to keep user_zero_stages (#3473)

Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>
Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>
Co-authored-by: Michael Wyatt <michaelwyatt@microsoft.com>

lekurile committed:
Update Inference Engine checkpoint loading + meta tensor assertions (#2940)

Co-authored-by: Ammar Ahmad Awan <ammar.awan@microsoft.com>

sywangyi committed:
fix regression in shard checkpoint loading in AutoTP Path caused by qkv_copy() is deleted and add UT case for shard checkpoint loading in AutoTP (#3457)

* add UT case for shard checkpoint loading in AutoTP

Signed-off-by: Wang, Yi A <yi.a.wang@intel.com>

* autoTP path also support shard loading

Signed-off-by: Wang, Yi A <yi.a.wang@intel.com>

---------

Signed-off-by: Wang, Yi A <yi.a.wang@intel.com>

ftian1 committed:
Add snip_momentum structured pruning which can support higher sparse ratio with minor accuracy loss (#3300)

Signed-off-by: Tian, Feng <feng.tian@intel.com>
Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

goodship1 committed:
Update README.md (#3504)

Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>

lekurile committed:
Hybrid Engine Fix Llama (#3505)

Co-authored-by: Connor Holmes <connorholmes@microsoft.com>
Co-authored-by: Jeff Rasley <jerasley@microsoft.com>
Co-authored-by: Ammar Ahmad Awan <ammar.awan@microsoft.com>

digger-yu committed:
fix spelling error with deepspeed/runtime/ (#3509)

Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>

molly-smith committed:
Skip autoTP if tp_size is 1 (#3449)

Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>
Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

jomayeri committed:
Changing monitor loss to aggregate loss over gradient accumulation steps (#3428)

* Changing monitor loss to aggregate loss over gas.

* Adding self.losses to engine constructor.

---------

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

digger-yu committed:
change actions/checkout@v2 to v3 (#3526)

digger-yu committed:
fix typo with docs/ (#3523)

* fix spelling error with deepspeed/runtime/

* fix typo docs/

---------

Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>

goodship1 committed:
Doc updates (#3520)

* Update pytorch-profiler.md

* Update one-cycle.md

---------

Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>

mrwyattii committed:
fix if statement for self.mpu (#3497)

mmhab committed:
Fix passing offload_optimizer_config instead of offload_param_config to (#3420)

DeepSpeedZeRoOffload

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>
Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>

YizhouZ committed:
Fix broadcast error on multi-node training with ZeroStage3 and TensorParallel=2 (#2999)

* * try to fix broadcast error on multi-node training with ZeroStage3 and TensorParallel=2

* * fix format error

* * fix format issue

* * add TODO for integrated testing of TP and ZeRO 1/2/3

* fix default pg error

---------

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

HeyangQin committed:
share inflight registry between PartitionedParameterCoordinators (#3462)

* share inflight registry between PartitionedParameterCoordinator

* bound registry to model

* make InflightParamRegistry standalone

* fix format

---------

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

jomayeri committed:
Syncing FusedAdam with new Apex features (#3434)

* Updating fused adam with new Apex bf16 support.

* Removing capturable and master weight configs.

* resolving pr comments

---------

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

digger-yu committed:
fix typo in comments with deepspeed/  (#3537)

* fix spelling error with deepspeed/runtime/

* fix typo docs/

* fix typo in comments with deepspeed/

---------

Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>
Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>

rraminen committed:
[ROCm] Hip headers fix (#3532)

* Add cg headers hipification

* Exclude including cuda_bf16.h on ROCm

* Merge

* Retricting including cuda_bf16.h with BF16_AVAILABLE var

---------

Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>

delock committed:
[CPU] Support Intel CPU inference (#3041)

* add fallback path for kernels used in megatron

* temporary numactl WA for SPR 56core

* adapt core allocation according to number of ranks

* add switch to turn on numactl

* detect number of cores on the system

* allow select a subset of the cores on the system to bind

* remove unneeded changes

* add ccl backend

* change nccl to ccl

* remove unused code

* add comm/ccl to ops

* initial ccl comm support

* first broadcast case passed

* add CCL_Backend to DeepSpeed

* support comm timer for CPU

* support barrier for comm backend

* support specify master address from deepspeed command line

* support pytorch 2.0

* remove 'block' from api

* Tweak for debug

Signed-off-by: Cao, Zhong Z <zhong.z.cao@intel.com>

* Remove unecessary directory

Signed-off-by: Cao, Zhong Z <zhong.z.cao@intel.com>

* Add bf16 kernel support for inference

* Add temporary torch implement for cpu inference

* Add softmax ops cpu fallback for inference

* bind cores to numa domain as well

* merge latest change in gma/numactl

* initial bf16 kernel support with fallback path

* initial fallback path for bloom kernel injection

* fix softmax attn mask

* check KMP_AFFINITY to avoid conflict with numactl

* New CCLBackend which utilize TorchBackend for initialization

* rollback last change because there is result error

* fix bloom injection policy TP could not work issue.

injection_policy={BloomBlock: ("self_attention.dense", "mlp.dense_4h_to_h")}

* Use TorchBackend to initialize CCLBackend, make behavior consistent

* remove comm under deepspeed/ops

* add license header

* code clean up

* fix format issue

* remove magic number in main address

* add caching support but not turn on by default

* change name of inference_cuda_module to inference_module

* Check for is_synchronized_device in accelerator before get Event

* fix typo

* Fix fallback path of softmax kernel on CUDA device for BF16 data type, because CUDA tril does not support BF16 datatype, enforce fp32 data type

* add cpu backend files

* change CPU_Accelerator op_builder_dir

* remove cpu_kernel_path

* using CPU_Accelerator on non-cuda device

* fix deepspeed.op_builder => deepspeed.ops.op_builder

* add alias for num_gpus: num_accelerators

* allow loading cpu_builder in build stage

* Assume cuda available if torch not installed

* add oneccl_binding_pt to requirements

* move oneccl-binding-pt to seperate requiremetns-cpu.txt

* add missing file

* use dependency_links in setuptools.setup() call for additional dependency links

* install oneccl_bind_pt in workflows

* change oneccl_bind_pt's version from 1.13 to 2.0

* use intel_exention_for_pytorch as indicator that CPU_Accelerator should be used

* Add indicator for Accelerator used

* change foo.c to foo.cpp

* exclude 'cpu' directory in CUDA op builder reflection

* add a cpu-inference workflow

* run cpu-inference workflow on self-hosted instance

* change cpu runs-on node to v100 node

* print out python version in workflow

* add verbose in pip command to understand oneccl_bind_pt install issue

* update cpu-inference workflow

* add a stage to detect instance instruction sets

* add back bf16 support for CPU inference

* enable autoTP for bloom

Signed-off-by: Wang, Yi A <yi.a.wang@intel.com>

* update workflow to detect cpu instruction sets

* temporary WA for Intel Extension for PyTorch AVX2 instructioon set detection

* change cpu-inference workflow machine to ubuntu-20.04

* add sharded checkpoint loading for AutoTP path to reduce the peak memory in initialization stage

Signed-off-by: Wang, Yi A <yi.a.wang@intel.com>

* enable policy for llama

* use a special build ipex to test avx2 detection fix

* fix format

* fix test fail issue

Signed-off-by: Wang, Yi A <yi.a.wang@intel.com>

* fix gptj sharded checkpoint loading problem

Signed-off-by: Wang, Yi A <yi.a.wang@intel.com>

* return a not implemented build in get_op_builder in cpu_backend

* support cpu device in tests

* use cpuinfo to extract number of CPUs

* use ~/tmp as transfomer cache rather than /blob/

* Add support for mpich launcher with prefer_deepspeed_comm

* add missing modification in accelerator

* enable IMPI launcher

* remove unused file and fix formatting

* clean up ccl.cpp

* Less confusing error message when certin op builder are not implemented

* Fix license header

* Add license header

* add license headers

* add license header

* fix cuda specific code in test

* update CPU workflow

* use numactl to bind to core

* allow bind_cores_to_rank in multi-node impi runner

* fix format error

* Remove InferenceBuilder

* fix format error in numa.py

* check whether op is in installed ops in ds_report.py

* allow override accelerator with DS_ACCELERATOR='cuda','cpu' or 'xpu'

* lazy init class_dict in CUDA_Accelerator to avoid cyclic initialization of CUDA_Accelerator

* put short path in the beginning in real_accelerator.py

* device_count return number of NUMA nodes

* fix typo

* install numactl in cpu workflow

* Follow comments

* Better implementation of device_count() and current_device()

* remove dependency_link for Intel Extension for DeepSpeed

* use check is_synchronized_device in timer only once

* remove env mapping WA in cpu_accelerator

* fix duplicate definition

* fix format error

* refine ccl backend selection

* move comments to the right place

* remove prefer_deepspeed_comm, use CCLBackend by default

* refractor fallback path

* Fix execution failure in kernel injection path

* do not refractory kernel injection fallback path in  residual_add because it contains function call with side-effect

* guard residual_add fallback path with environ DS_KI_FALLBACK=True

* fix format error

* add test for allreduce on CPU workflow

* fix format error

* Fallback to TorchBackend if CCLBackend kernel are not implemented

* Update Intel Extension for Pytorch installation link

* Don't specify version number of Intel Extension for PyTorch

* install oneCCL for CCLBackend

* fix link path for CPU comm kernels

* fix source oneCCL environment

* source oneCCL env before run UT

* Give more specific instruction when CCL_ROOT not defined

---------

Signed-off-by: Cao, Zhong Z <zhong.z.cao@intel.com>
Signed-off-by: Wang, Yi A <yi.a.wang@intel.com>
Co-authored-by: sdp <sdp@aia-sdp-spr-108864.jf.intel.com>
Co-authored-by: Cao, Zhong Z <zhong.z.cao@intel.com>
Co-authored-by: Zhenhuan Chen <zhenhuan.chen@intel.com>
Co-authored-by: baodii <di.bao@intel.com>
Co-authored-by: Wang, Yi A <yi.a.wang@intel.com>
Co-authored-by: jianan-gu <jianan.gu@intel.com>
Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>
Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>

tjruwase committed:
Clone tensors to avoid torch.save bloat (#3348)

* Clone tensors to avoid torch.save bloat

* Adddocs

* Fix clang-formatting

* Update docs/code-docs/source/model-checkpointing.rst

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

* Update deepspeed/checkpoint/utils.py

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

* Update deepspeed/checkpoint/utils.py

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

* Fix url

* url fix

* Tweak docs

---------

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>
Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>

rraminen committed:
Fix (#3527)

Co-authored-by: Logan Adams <114770087+loadams@users.noreply.github.com>
Co-authored-by: Olatunji Ruwase <olruwase@microsoft.com>

inkcherry committed:
fix_typo (#3559)
